pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -362,13 +362,20 @@ private void initPolicesCache(SystemTopicClient.Reader<PulsarEvent> reader, Comp
                 policyCacheInitMap.computeIfPresent(
                         reader.getSystemTopic().getTopicName().getNamespaceObject(), (k, v) -> true);
                 // replay policy message
-                policiesCache.forEach(((topicName, topicPolicies) -> {
-                    if (listeners.get(topicName) != null) {
-                        for (TopicPolicyListener<TopicPolicies> listener : listeners.get(topicName)) {
-                            listener.onUpdate(topicPolicies);
+                try {
+                    policiesCache.forEach(((topicName, topicPolicies) -> {
+                        if (listeners.get(topicName) != null) {
+                            for (TopicPolicyListener<TopicPolicies> listener : listeners.get(topicName)) {
+                                listener.onUpdate(topicPolicies);
+                            }
                         }
-                    }
-                }));
+                    }));
+                } catch (Throwable e) {
+                    log.error("[{}] Failed to replay policy message.", reader.getSystemTopic().getTopicName(), e);
+                    future.completeExceptionally(e);
+                    cleanCacheAndCloseReader(reader.getSystemTopic().getTopicName().getNamespaceObject(), false);
```

Comment:
We should not close the reader here because the caller will close it.

Reply:
yes, right. We shouldn't close the reader here

Reply:
cleanCacheAndCloseReader had delete from here.